src/Native/LDANative/lda_engine.cpp

```diff
@@ -0,0 +1,1103 @@
+        for (int k = 0; k < K_; ++k)
+        {
+            proportion[k] = beta_ / (global_summary_row_[k] + beta_sum_);
+            beta_mass_ += proportion[k];
+        }
+        alias_rng_int_.SetProportionMass(proportion, beta_mass_, beta_k_v_, &beta_height_, samplers_[0]->rng());
+
+        // 2,  build alias table for the sparse term
+        for (int thread_id = 0; thread_id < num_threads_; ++thread_id)
+        {
+            LightDocSampler &sampler = *(samplers_[thread_id]);
+            sampler.build_alias_table(word_range_for_each_thread_[thread_id], word_range_for_each_thread_[thread_id + 1], thread_id);
+        }
+        CheckFunction(0, tmDebug, "build alisa table", false);
+    }
+
+    void LdaEngine::Train(const char* pTrainOutput)
+    {
+        std::vector<std::thread> threads(num_threads_);
+        atomic_stats_->thread_counter_ = 0;
+
+        for (auto& thr : threads) {
+            thr = std::thread(&LdaEngine::Training_Thread, this);
+        }
+
+        printf("started training with %d threads\n", num_threads_);
+        for (auto& thr : threads) {
+            thr.join();
+        }
+
+        if (pTrainOutput)
+        {
+            DumpDocTopicTable(pTrainOutput);
+        }
+    }
+
+    void LdaEngine::Test(int32_t burnin_iter, float* pLoglikelihood)
+    {
+        std::vector<std::thread> threads(num_threads_);
+        atomic_stats_->thread_counter_ = 0;
+        burnin_iterations_ = burnin_iter;
+
+        likelihood_in_iter_ = new float[burnin_iterations_];
+        for (int i = 0; i < burnin_iterations_; i++)
+        {
+            likelihood_in_iter_[i] = 0.0;
+        }
+
+        for (auto& thr : threads) {
+            thr = std::thread(&LdaEngine::Testing_Thread, this);
+        }
+
+        printf("started testing with %d threads\n", num_threads_);
+
+        for (auto& thr : threads) {
+            thr.join();
+        }
+
+        //get the loglikelihood of each burn in iteration
+        for (int i = 0; i < burnin_iterations_; i++)
+        {
+            pLoglikelihood[i] = likelihood_in_iter_[i]; //just set an arbitary value here for later update
+        }
+    }
+
+    void LdaEngine::CheckFunction(int thread_id, CTimer &tmDebug, const char* msg, bool waitBarrier)
+    {
+        /*if (thread_id == 0)
+        {
+        sprintf(tmDebug.m_szMessage, msg);
+        tmDebug.InnerTag();
+        system("pause");
+        }
+
+        if (waitBarrier)
+        {
+        process_barrier_->wait();
+        }*/
+    }
+
+    void LdaEngine::Training_Thread()
+    {
+        CTimer tmDebug(true);
+
+        int thread_id = atomic_stats_->thread_counter_++;
+        std::vector<std::pair<int, double>> llcontainer;
+        // Set core affinity which helps performance improvement
+#ifdef _MSC_VER
+        long long maskLL = 0;
+        maskLL |= (1LL << (thread_id));
+        DWORD_PTR mask = maskLL;
+        SetThreadAffinityMask(GetCurrentThread(), mask);
+#elif !defined(__APPLE__)
```

Comment:
>!defined(__APPLE__) [](start = 6, length = 19)

what about __APPLE__?

Reply:
will create issue to track this separately